torch/csrc/autograd/record_function_ops.cpp

```diff
@@ -14,42 +14,44 @@ namespace torch {
 namespace autograd {
 namespace profiler {
 
-// Creates a new profiling scope using RecordFunction and invokes its starting
-// callbacks.
-at::Tensor record_function_enter(const std::string& name) {
-  auto rec = std::make_unique<at::RecordFunction>(at::RecordScope::USER_SCOPE);
-  rec->before(name);
-  return at::cpp_custom_type_hack::create(std::move(rec), at::TensorOptions());
+void RecordFunctionHolder::enter(const std::string& name) {
+  record_function_ = std::make_unique<at::RecordFunction>(at::RecordScope::USER_SCOPE);
+  record_function_->before(name);
 }
 
-at::RecordFunction& getRecordFunctionFromTensor(const at::Tensor& handle) {
-  auto& rec = at::cpp_custom_type_hack::cast<at::RecordFunction>(handle);
-  return rec;
+void RecordFunctionHolder::exit() {
+  TORCH_CHECK(record_function_ != nullptr,
```

Comment:
We usually use TORCH_INTERNAL_ASSERT for bugs that are due to developer coding logic, and TORCH_CHECK for issues that are due to incorrect use of pytorch by the user. Since it doesn't appear that the user can get the application in this state, we should probably switch it to the former.

Also if it incurs a lot of overhead, we can switch to a debug check and it will be stripped out in non-debug builds.

Reply:
Got it